HW0/P6/P6.txt

```diff
@@ -1 +1,131 @@
-# Your answers here
+P6A
+----
+See below for sample output of running P6A.py
+It's clear from this output that not all jobs complete in exactly the same order as they begin. On each run, the "Hi" outputs all appear in the same expected order, but the "Bye" outputs at the conclusion of each operation appear in a completely different order. In fact, on different runs, the ordering of the "Bye" outputs is likewise completely different. This can be explained by the fact that despite the fact that the wait times across all concurrent threads are the same, some threads finish quicker that others (this is likely caused by the way the machine actually allocates resources to each thread).
+This could seriously affect our ability to parallelize our programs if some of these calls depended on the others to complete in the expected order or at some expected time. If this was the case, we could run into serious issues with consistency and expected behavior. 
+Take, for example, the naive money counting example from P5. If, instead of communcating, the cashiers in the second level of computation were simply placed later in the input list to a program like this, assuming that by the time it was theur turn to compute, the previous computations would have already been made. This expectation, if it turned out to be false (as is shown to be possible by the below "Hi"/"Bye" output) could easily lead cashiers later in the line to make summations with 0s, garbage values, or other values entirely if the order of the previous cashiers got jumbled.
+
+P6B
+----
+The trend is generally upwards, beginning below the 0.5 mark and tapering off at around the 4.0 mark. This makes sense, as with very short computation time, serial processing actually performs better (as will be explained shortly), while at long computation times, the parallelized solution wins out by no better than the number of processes permitted to run concurrently (in this case, 4). 
+It is possible for parallel threads to take longer than the serial executions. In this graph, we see that this happens around and below the 10^-5 mark of computation time. Thus, when the actual wait time per action is small enough, the overhead of distributing and recombining resources actually slows the parallelized solution down relative to the serialized one. This makes sense if we again consider the bags of coins problem in P5. There, we noticed that when the compute time at the lowest level was only 1 second, the parallelized solution spent more time handing out the bags to the cashiers than it would have taken to simply make a serial computation. In that problem, we found that at longer computation times, with the overhead of multithreading reduced, the parallel solution proved much more effective. This exact behavior can be seen through the P6.png graph included here.
```

Comment:
I like your examples relating back to P5. It is most clear when thinking about having 4 bags, handing 2 out to be computed, while you compute your 2 bags. That would take 1 hand out, 1 parallel compute, 1 communication back, and 1 communication to add the two sums, while adding the 4 numbers by themselves takes 3 seconds. The overhead starts to really get in the way of computing. However, it seems to still be the best strategy to parallelize when there are a larger amount of bags, so since the tradeoff occurs roughly when computing and parallizing are equal speeds, we can conclude from this exercise that the overhead to run in parallel takes about 10^(-5) seconds